PythonVisualizations/Stack.py

```diff
@@ -0,0 +1,318 @@
+    CELL_BORDER = 2
+    CELL_BORDER_COLOR = 'black'
+    STACK_X0 = 300
+    STACK_Y0 = 400
+
+    def __init__(self, size=MAX_SIZE, maxArgWidth=MAX_ARG_WIDTH, title="Stack", **kwargs):
+        super().__init__(title=title, maxArgWidth=maxArgWidth, **kwargs)
+        self.size = size
+        self.title = title
+        self.list = []
+        self.maxArgWidth = maxArgWidth
+        self.buttons = self.makeButtons()
+        self.display()
+        self.indexDisplay = self.createIndex(len(self.list))
+
+    def __str__(self):
+        return str(self.list)
+
+    # Create an index arrow to point at an indexed cell with an optional name label
+    def createIndex(self, index, name=None):
+        cell_coords = self.cellCoords(index)
+        cell_center = self.cellCenter(index)
+        x0 = self.STACK_X0 - self.CELL_WIDTH * 4 // 5
+        x1 = self.STACK_X0 - self.CELL_WIDTH * 3 // 10
+        y0 = y1 = cell_coords[1] + self.CELL_HEIGHT // 2
+        if not name:
+            label = "top" #labels the top of the stack "top" with the pointer arrow
+        else:
+            label = name
+
+        return self.drawArrow(
+            x0, y0, x1, y1, self.VARIABLE_COLOR, self.VARIABLE_FONT, name=label)
+
+    # draw the actual arrow
+    def drawArrow(
+            self, x0, y0, x1, y1, color, font, name=None):
+        arrow = self.canvas.create_line(
+            x0, y0, x1, y1, arrow="last", fill=color)
+        if name:
+            label = self.canvas.create_text(
+                x0 - self.CELL_WIDTH / 2, y0 + self.CELL_HEIGHT / 5, text=name, anchor=SW,
+                font=font, fill=color)
+
+        return (arrow, label) if name else (arrow,)
+
+    # STACK FUNCTIONALITY
+
+    def push(self, val):
+
+        self.startAnimations()
+        callEnviron = self.createCallEnvironment()
+
+        #don't move arrow up if the first cell is being filled because it is already pointing there
+        if len(self.list) >= 1:
+            self.moveItemsBy(self.indexDisplay, (0, - (self.CELL_HEIGHT)))
+
+        cellCoords = self.cellCoords(len(self.list))
+        cellCenter = self.cellCenter(len(self.list))
+
+        # create new cell and cell value display objects
+        cellCoords = add_vector(cellCoords, (0, 0, 0, self.CELL_BORDER))
+        toPositions = (cellCoords, cellCenter)
+
+        # determine the top left and bottom right positions
+        startPosition = [self.STACK_X0, 0, self.STACK_X0, 0]
+        startPosition = add_vector(startPosition, (0, 0, self.CELL_WIDTH, self.CELL_HEIGHT))
+        cellPair = self.createCellValue(startPosition, val)
+        callEnviron.add(cellPair)
+        self.moveItemsTo(cellPair, toPositions, steps=self.CELL_HEIGHT, sleepTime=0.01)
+
+        # add a new Drawable with the new value, color, and display objects
+        d = drawable(
+            val, self.canvas.itemconfigure(cellPair[0], 'fill')[-1], *cellPair)
+        self.list.append(d)
+        callEnviron.discard(cellPair)
```

Comment:
The `cellPair` created in line 83 is a tuple of canvas items.  When `.add()` is used to add that tuple to the `callEnviron` set, it adds a single element to the set (the tuple object).  The desired result is to add the two canvas items individually to the set.  That can be done with a loop like:
```
    for item in cellPair:
        callEnviron.add(item)
```
or with a set union operation like:
```
    callEnviron |= set(cellPair)
```

A parallel change is needed on line 91 where the items are removed from the call environment (because they should now stick around with the other stack items).  So either:
```
    for item in cellPair:
        callEnviron.discard(item)
```
or with a set exclusive-or operation like:
```
    callEnviron ^= set(cellPair)
```
or with a set difference operation like:
```
    callEnviron -= set(cellPair)
```
Note: we won't be able to see the effects of putting the `cellPair` in the `callEnviron` as a tuple until the Stop button is added.  The canvas items will remain on screen and don't need to be cleaned up until the user can interrupt the animation of the movement.